File: PHP/Avanzado/web_scraper/scraper.php
<?php

require 'vendor/autoload.php';

use GuzzleHttp\Client;
use Symfony\Component\DomCrawler\Crawler;

function scrape(string $url, string $selector): array
{
    $client = new Client(['headers' => ['User-Agent' => 'Mozilla/5.0']]);
    $html   = (string)$client->get($url)->getBody();
    $crawler = new Crawler($html);

    return $crawler->filter($selector)->each(fn(Crawler $nodo) => [
        'texto' => trim($nodo->text()),
        'href'  => $nodo->attr('href'),
    ]);
}

function guardar_json(array $datos, string $archivo): void
{
    file_put_contents($archivo, json_encode($datos, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
    echo "Guardado en $archivo\n";
}

function guardar_csv(array $datos, string $archivo): void
{
    if (empty($datos)) return;
    $fp = fopen($archivo, 'w');
    fputcsv($fp, array_keys($datos[0]));
    foreach ($datos as $fila) fputcsv($fp, $fila);
    fclose($fp);
    echo "Guardado en $archivo\n";
}

echo 'URL: ';
$url = trim(fgets(STDIN));
echo 'Selector CSS (ej: a, h2, .clase): ';
$selector = trim(fgets(STDIN));

$datos = scrape($url, $selector);
echo "Encontrados: " . count($datos) . " elementos\n";
foreach (array_slice($datos, 0, 5) as $d) {
    echo $d['texto'] . "\n";
}

guardar_json($datos, 'resultado.json');
guardar_csv($datos, 'resultado.csv');

File: Ruby/Intermedio/scraper_nokogiri/scraper.rb
require 'nokogiri'
require 'open-uri'
require 'json'
require 'csv'

def scrape(url, selector)
  doc = Nokogiri::HTML(URI.open(url, 'User-Agent' => 'Mozilla/5.0'))
  doc.css(selector).map do |el|
    { texto: el.text.strip, href: el['href'] }
  end
end

def guardar_json(datos, archivo)
  File.write(archivo, JSON.pretty_generate(datos))
  puts "Guardado en #{archivo}"
end

def guardar_csv(datos, archivo)
  return if datos.empty?

  CSV.open(archivo, 'w', headers: datos.first.keys, write_headers: true) do |csv|
    datos.each { |row| csv << row.values }
  end
  puts "Guardado en #{archivo}"
end

print 'URL: '
url = gets.chomp
print 'Selector CSS (ej: a, h2, .clase): '
selector = gets.chomp

datos = scrape(url, selector)
puts "Encontrados: #{datos.length} elementos"
datos.first(5).each { |d| puts d }

guardar_json(datos, 'resultado.json')
guardar_csv(datos, 'resultado.csv')
